src/fdb5/database/Key.h

```diff
@@ -159,9 +153,73 @@ class Key {
 
     eckit::StringDict keys_;
     eckit::StringList names_;
+};
+
+
+//----------------------------------------------------------------------------------------------------------------------
+
+class Key : public BaseKey {
+
+public: // methods
+
+    explicit Key();
+    explicit Key(eckit::Stream &);
+    explicit Key(const eckit::StringDict &keys);
+    Key(std::initializer_list<std::pair<const std::string, std::string>>);
+
+    static Key parseString(const std::string& s);
+
+    friend eckit::Stream& operator>>(eckit::Stream& s, Key& x) {
+        x = Key(s);
+        return s;
+    }
+
+private: // members
+
+    std::string canonicalise(const std::string& keyword, const std::string& value) const override;
+    std::string type(const std::string& keyword) const override;
+
+};
+
+//----------------------------------------------------------------------------------------------------------------------
+
+class TypedKey : public BaseKey {
+
+public: // methods
+
+    explicit TypedKey(const Key& key, const std::shared_ptr<TypesRegistry> reg);
+    explicit TypedKey(const std::shared_ptr<TypesRegistry> reg);
+    explicit TypedKey(eckit::Stream &, const std::shared_ptr<TypesRegistry> reg);
+    explicit TypedKey(const std::string &keys, const Rule* rule);
+    explicit TypedKey(const eckit::StringDict &keys, const std::shared_ptr<TypesRegistry> reg);
+    TypedKey(std::initializer_list<std::pair<const std::string, std::string>>, const std::shared_ptr<TypesRegistry> reg);
+
+    static TypedKey parseString(const std::string&, const std::shared_ptr<TypesRegistry> reg);
+
+    Key canonical() const;
+
+    /// @throws When "other" doesn't contain all the keys of "this"
+    void validateKeys(const BaseKey& other, bool checkAlsoValues = false) const;
+
+    friend eckit::Stream& operator>>(eckit::Stream& s, TypedKey& x) {
+        x = TypedKey(s, nullptr);
+        return s;
+    }
+
+    // Registry is needed before we can stringise/canonicalise.
+    void registry(const std::shared_ptr<TypesRegistry> reg);
+    [[ nodiscard ]]
+    const TypesRegistry& registry() const;
+    const void* reg() const;
+
+private: // members
+
+    //TODO add unit test for each type
+    std::string canonicalise(const std::string& keyword, const std::string& value) const override;
+    std::string type(const std::string& keyword) const override;
 
     std::shared_ptr<TypesRegistry> registry_;
```

Comment:
non-const `registry_` member looks good, 
no reason why not `std::shared_ptr<const TypesRegistry>` ?

also, all `const` qualification of class methods for `req` param are redundant